entity-service-client/src/main/java/org/hypertrace/entity/query/service/client/EntityLabelsCachingClient.java

```diff
@@ -0,0 +1,216 @@
+                                               Map<String, String> headers,
+                                               String tenantId) {
+    try {
+      return entityIdsToLabelsCache.get(
+          new EntityCacheKey<>(new EntityTypeAndId(type, id, idColumnName, labelsColumnName),
+              tenantId,
+              headers
+          )
+      );
+    } catch (ExecutionException ex) {
+      throw new RuntimeException("EntityLabelsCachingClient exception while getting labels for an entity", ex);
+    }
+  }
+
+  @Override
+  public void refreshEntityLabelsForEntity(String idColumnName,
+                                           String labelsColumnName,
+                                           String id,
+                                           String type,
+                                           Map<String, String> headers,
+                                           String tenantId) {
+    entityIdsToLabelsCache.refresh(
+        new EntityCacheKey<>(new EntityTypeAndId(type, id, idColumnName, labelsColumnName),
+            tenantId,
+            headers
+        )
+    );
+  }
+
+  @Override
+  public Map<String, List<String>> getEntityLabelsForEntities(String idColumnName,
+                                                              String labelsColumnName,
+                                                              List<String> ids,
+                                                              String type,
+                                                              Map<String, String> headers,
+                                                              String tenantId) {
+    List<EntityCacheKey<EntityTypeAndId>> entityCacheKeys = ids.stream()
+        .map(entityId -> new EntityTypeAndId(type, entityId, idColumnName, labelsColumnName))
+        .map(entityTypeAndId -> new EntityCacheKey<>(entityTypeAndId, tenantId, headers))
+        .collect(Collectors.toList());
+    try {
+      Map<EntityCacheKey<EntityTypeAndId>, List<String>> entityCacheKeyToLabelsMap =  entityIdsToLabelsCache.getAll(entityCacheKeys);
+      return entityCacheKeyToLabelsMap.entrySet().stream()
+          .collect(
+              Collectors.toMap(entry -> entry.getKey().getDataKey().getId(), Entry::getValue)
+          );
+    } catch (ExecutionException ex) {
+      throw new RuntimeException("EntityLabelsCachingClient exception while getting labels for entities", ex);
+    }
+  }
+
+  private Map<EntityCacheKey<EntityTypeAndId>, List<String>> loadEntityLabelsForKeys(Iterable<? extends EntityCacheKey<EntityTypeAndId>> keys) {
+    Iterator<? extends EntityCacheKey<EntityTypeAndId>> iter = keys.iterator();
+    if (!iter.hasNext()) {
+      return Map.of();
+    }
+
+    List<EntityCacheKey<EntityTypeAndId>> entityIdKeys = new ArrayList<>();
+    while (iter.hasNext()) {
+      entityIdKeys.add(iter.next());
+    }
+
+    // The assumption we make is that the all the entity id keys are of the same entity type and
```

Comment:
If we took those in as params, we could unwind that assumption. But really the only way to completely avoid it would be to store it as a tree, where the top level key is the context and then below that we store entities. Otherwise, loadAll could always be invoked with multiple different contexts.